File: CS/14_Page/Pagination.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Text;
using System.Windows.Forms;
using Spire.Pdf;
using Spire.Pdf.Graphics;


namespace Pagination
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            // Create a new instance of a PDF document
            PdfDocument doc = new PdfDocument();

            // Set the margin for the document
            PdfUnitConvertor unitCvtr = new PdfUnitConvertor();
            PdfMargins margin = new PdfMargins();
            margin.Top = unitCvtr.ConvertUnits(2.54f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Bottom = margin.Top;
            margin.Left = unitCvtr.ConvertUnits(3.17f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Right = margin.Left;

            // Draw the cover page using specified margin and add it to the document
            DrawCover(doc, doc.Sections.Add(), margin);

            // Draw the content page using specified margin and add it to the document
            DrawContent(doc, doc.Sections.Add(), margin);

            // Draw the page number on the second section of the document using specified margin,
            // starting from page 1 and counting the total number of pages in the second section
            DrawPageNumber(doc.Sections[1], margin, 1, doc.Sections[1].Pages.Count);

            // Save the document to a file named "Pagination.pdf"
            doc.SaveToFile("Pagination.pdf");

            // Close the document
            doc.Close();

            //Launch the Pdf file
            PDFDocumentViewer("Pagination.pdf");
        }

        private void DrawCover(PdfDocument pdf, PdfSection section, PdfMargins margin)
        {
            // Set the page size of the section to A4
            section.PageSettings.Size = PdfPageSize.A4;

            // Set the margins of the section to zero
            section.PageSettings.Margins.All = 0;

            // Add a new page to the section
            PdfPageBase page = section.Pages.Add();

            // Create a new instance of PdfPageLabels for the document
            pdf.PageLabels = new PdfPageLabels();

            // Add page labels to the document starting from index 0 with lowercase Roman numerals and prefix "cover "
            pdf.PageLabels.AddRange(0, PdfPageLabels.Lowercase_Roman_Numerals_Style, "cover ");

            // Call a method to draw the header and footer on the page
            DrawPageHeaderAndFooter(page, margin, true);

            // Set up colors, fonts, and text format for the content
            PdfBrush brush1 = PdfBrushes.LightGray;
            PdfBrush brush2 = PdfBrushes.Blue;
            PdfTrueTypeFont font1 = new PdfTrueTypeFont(new Font("Arial", 8f));
            PdfStringFormat format = new PdfStringFormat();
            format.MeasureTrailingSpaces = true;
            String text1 = "(All text and picture from ";
            String text2 = "Wikipedia";
            String text3 = ", the free encyclopedia)";
            float x = 0, y = 10;

            // Adjust the starting position of drawing based on the margin
            x = x + margin.Left;
            y = y + margin.Top;

            // Draw the content strings on the page using the specified fonts, brushes, and format
            page.Canvas.DrawString(text1, font1, brush1, x, y, format);
            x = x + font1.MeasureString(text1, format).Width;
            page.Canvas.DrawString(text2, font1, brush2, x, y, format);
            x = x + font1.MeasureString(text2, format).Width;
            page.Canvas.DrawString(text3, font1, brush1, x, y, format);

            // Set up colors, image, and other parameters for the cover
            PdfBrush brush3 = PdfBrushes.Black;
            PdfBrush brush4 = new PdfSolidBrush(new PdfRGBColor(0xf9, 0xf9, 0xf9));
            PdfImage image = PdfImage.FromFile(@"..\..\..\..\..\..\Data\SciencePersonificationBoston.jpg");
            String text = "Personification of \"Science\" in front of the Boston Public Library";
            float r = image.PhysicalDimension.Height / image.Height;
            PdfPen pen = new PdfPen(brush1, r);
            SizeF size = font1.MeasureString(text, image.PhysicalDimension.Width - 2);
            PdfTemplate template = new PdfTemplate(image.PhysicalDimension.Width + 4 * r + 4, image.PhysicalDimension.Height + 4 * r + 7 + size.Height);

            // Draw a rectangle with border and fill on the template
            template.Graphics.DrawRectangle(pen, brush4, 0, 0, template.Width, template.Height);

            // Adjust the starting position of drawing based on the margin and radius
            x = y = r + 2;

            // Draw a rectangle on the template with a brush
            template.Graphics.DrawRectangle(brush1, x, y,
                image.PhysicalDimension.Width + 2 * r,
                image.PhysicalDimension.Height + 2 * r);

            // Adjust the starting position of drawing within the rectangle
            x = y = x + r;

            // Draw the image on the template
            template.Graphics.DrawImage(image, x, y);

            // Adjust the starting position of drawing for the text below the image
            x = x - 1;
            y = y + image.PhysicalDimension.Height + r + 2;

            // Draw the text on the template using specified font, brush, and rectangle
            template.Graphics.DrawString(text, font1, brush3,
                new RectangleF(new PointF(x, y), size));

            // Calculate the positioning of the template and draw it on the page canvas
            float x1 = (page.Canvas.ClientSize.Width - template.Width) / 2;
            float y1 = (page.Canvas.ClientSize.Height - margin.Top - margin.Bottom) * (1 - 0.618f) - template.Height / 2 + margin.Top;
            template.Draw(page.Canvas, x1, y1);

            // Set up alignment and font for the title text
            format.Alignment = PdfTextAlignment.Center;
            PdfTrueTypeFont font2 = new PdfTrueTypeFont(new Font("Arial", 24f, FontStyle.Bold));

            // Calculate the positioning of the title text and draw it on the page canvas
            x = page.Canvas.ClientSize.Width / 2;
            y = y1 + template.Height + 10;
            page.Canvas.DrawString("Science History and Etymology", font2, brush3, x, y, format);
        }

        private void DrawContent(PdfDocument pdf, PdfSection section, PdfMargins margin)
        {
            // Set the page size of the section to A4
            section.PageSettings.Size = PdfPageSize.A4;

            // Set the margins of the section to zero
            section.PageSettings.Margins.All = 0;

            // Add a new page to the section
            PdfPageBase page = section.Pages.Add();

            // Add page labels to the document starting from index 1 with decimal Arabic numerals and prefix "page "
            pdf.PageLabels.AddRange(1, PdfPageLabels.Decimal_Arabic_Numerals_Style, "page ");

            // Call a method to draw the header and footer on the page without cover information
            DrawPageHeaderAndFooter(page, margin, false);

            // Set up initial positions and dimensions
            float x = margin.Left;
            float y = margin.Top + 8;
            float width = page.Canvas.ClientSize.Width - margin.Left - margin.Right;

            // Set up font, brush, and pen for the title
            PdfTrueTypeFont font1 = new PdfTrueTypeFont(new Font("Arial", 16f), true);
            PdfBrush brush1 = PdfBrushes.Black;
            PdfPen pen1 = new PdfPen(brush1, 0.75f);

            // Draw the title text on the page canvas
            page.Canvas.DrawString("Science History and Etymology", font1, brush1, x, y);

            // Adjust the position and draw a horizontal line below the title
            y = y + font1.MeasureString("Science History and Etymology").Height + 6;
            page.Canvas.DrawLine(pen1, x, y, page.Canvas.ClientSize.Width - margin.Right, y);
            y = y + 1.75f;

            // Set up content strings and format options
            String content = "Main articles: History of science and Scientific revolution\r\nWhile descriptions of disciplined empirical investigations of the natural world exist from times at least as early as classical antiquity (for example, by Aristotle and Pliny the Elder), and scientific methods have been employed since the Middle Ages (for example, by Alhazen and Roger Bacon), the dawn of modern science is generally traced back to the early modern period during what is known as the Scientific Revolution of the 16th and 17th centuries. This period was marked by a new way of studying the natural world, by methodical experimentation aimed at defining \"laws of nature\" while avoiding concerns with metaphysical concerns such as Aristotle's theory of causation.\r\n\"Rapid accumulation of knowledge, which has characterized the development of science since the 17th century, had never occurred before that time. The new kind of scientific activity emerged only in a few countries of Western Europe, and it was restricted to that small area for about two hundred years. (Since the 19th century, scientific knowledge has been assimilated by the rest of the world).\" ¡ª Joseph Ben¨CDavid, 1971.\r\nThis modern science developed from an older and broader enterprise. The word \"science\" is from Old French, and in turn from Latin scientia which was one of several words for \"knowledge\" in that language. In philosophical contexts, scientia and \"science\" were used to translate the Greek word epistem¨¥, which had acquired a specific definition in Greek philosophy, especially Aristotle, as a type of reliable knowledge which is built up logically from strong premises, and can be communicated and taught. In contrast to modern science, Aristotle's influential emphasis was upon the \"theoretical\" steps of deducing universal rules from raw data, and did not treat the gathering of experience and raw data as part of science itself.\r\nFrom the Middle Ages to the Enlightenment, science or scientia continued to be used in this broad sense, which was still common until the 20th century. \"Science\" therefore had the same sort of very broad meaning that philosophy had at that time. In other Latin influenced languages, including French, Spanish, Portuguese, and Italian, the word corresponding to science also carried this meaning.\r\nPrior to the 18th century, the preferred term for the study of nature among English speakers was \"natural philosophy\", while other philosophical disciplines (e.g., logic, metaphysics, epistemology, ethics and aesthetics) were typically referred to as \"moral philosophy\". (Today, \"moral philosophy\" is more-or-less synonymous with \"ethics\".) Science only became more strongly associated with natural philosophy than other sciences gradually with the strong promotion of the importance of experimental scientific method, by people such as Francis Bacon. With Bacon, begins a more widespread and open criticism of Aristotle's influence which had emphasized theorizing and did not treat raw data collection as part of science itself. An opposed position became common: that what is critical to science at its best is methodical collecting of clear and useful raw data, something which is easier to do in some fields than others.\r\nThe word \"science\" in English was still however used in the 17th century to refer to the Aristotelian concept of knowledge which was secure enough to be used as a prescription for exactly how to accomplish a specific task. With respect to the transitional usage of the term \"natural philosophy\" in this period, the philosopher John Locke wrote in 1690 that \"natural philosophy is not capable of being made a science\". However, it may be that Locke was not using the word 'science' in the modern sense, but suggesting that 'natural philosophy' could not be deduced in the same way as mathematics and logic.\r\nLocke's assertion notwithstanding, by the early 19th century natural philosophy had begun to separate from philosophy, though it often retained a very broad meaning. In many cases, science continued to stand for reliable knowledge about any topic, in the same way it is still used today in the broad sense (see the introduction to this article) in modern terms such as library science, political science, and computer science. In the more narrow sense of science, as natural philosophy became linked to an expanding set of well-defined laws (beginning with Galileo's laws, Kepler's laws, and Newton's laws for motion), it became more popular to refer to natural philosophy as natural science. Over the course of the 19th century, moreover, there was an increased tendency to associate science with study of the natural world (that is, the non-human world). This move sometimes left the study of human thought and society (what would come to be called social science) in a linguistic limbo by the end of the century and into the next.\r\nThrough the 19th century, many English speakers were increasingly differentiating science (i.e., the natural sciences) from all other forms of knowledge in a variety of ways. The now-familiar expression ¡°scientific method,¡± which refers to the prescriptive part of how to make discoveries in natural philosophy, was almost unused until then, but became widespread after the 1870s, though there was rarely total agreement about just what it entailed. The word \"scientist,\" meant to refer to a systematically working natural philosopher, (as opposed to an intuitive or empirically minded one) was coined in 1833 by William Whewell. Discussion of scientists as a special group of people who did science, even if their attributes were up for debate, grew in the last half of the 19th century. Whatever people actually meant by these terms at first, they ultimately depicted science, in the narrow sense of the habitual use of the scientific method and the knowledge derived from it, as something deeply distinguished from all other realms of human endeavor.\r\nBy the 20th century, the modern notion of science as a special kind of knowledge about the world, practiced by a distinct group and pursued through a unique method, was essentially in place. It was used to give legitimacy to a variety of fields through such titles as \"scientific\" medicine, engineering, advertising, or motherhood. Over the 20th century, links between science and technology also grew increasingly strong. As Martin Rees explains, progress in scientific understanding and technology have been synergistic and vital to one another.\r\nRichard Feynman described science in the following way for his students: \"The principle of science, the definition, almost, is the following: The test of all knowledge is experiment. Experiment is the sole judge of scientific 'truth'. But what is the source of knowledge? Where do the laws that are to be tested come from? Experiment, itself, helps to produce these laws, in the sense that it gives us hints. But also needed is imagination to create from these hints the great generalizations ¡ª to guess at the wonderful, simple, but very strange patterns beneath them all, and then to experiment to check again whether we have made the right guess.\" Feynman also observed, \"...there is an expanding frontier of ignorance...things must be learned only to be unlearned again or, more likely, to be corrected.\"";
            String[] lines = content.Split(new char[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
            PdfTrueTypeFont font2 = new PdfTrueTypeFont(new Font("Arial", 10f, FontStyle.Italic), true);
            PdfStringFormat format1 = new PdfStringFormat();
            format1.MeasureTrailingSpaces = true;
            format1.LineSpacing = font2.Height * 1.5f;
            format1.ParagraphIndent = font2.MeasureString("\t", format1).Width;

            // Adjust the starting position for the first line of content
            y = y + font2.Height * 0.5f;

            // Measure and draw the first line of content using the specified font, brush, rectangle, and format
            SizeF size = font2.MeasureString(lines[0], width, format1);
            page.Canvas.DrawString(lines[0], font2, brush1,
                new RectangleF(new PointF(x, y), size), format1);

            // Adjust the position for the next line
            y = y + size.Height;

            // Set up font and format options for the remaining lines of content
            PdfTrueTypeFont font3 = new PdfTrueTypeFont(new Font("Arial", 10f), true);
            PdfStringFormat format2 = new PdfStringFormat();
            format2.LineSpacing = font3.Height * 1.4f;
            format2.MeasureTrailingSpaces = true;

            // Measure and draw the second line of content using the specified font, brush, rectangle, and format
            size = font3.MeasureString(lines[1], width, format2);
            page.Canvas.DrawString(lines[1], font3, brush1,
                new RectangleF(new PointF(x, y), size), format2);

            // Adjust the position for the next line
            y = y + size.Height;

            // Adjust the position and indentation for the third line of content
            y = y + font3.Height * 0.75f;
            float indent = font3.MeasureString("\t\t", format2).Width;
            float x1 = x + indent;

            // Measure and draw the third line of content using the specified font, brush, rectangle, and format
            size = font3.MeasureString(lines[2], width - indent, format2);
            page.Canvas.DrawString(lines[2], font3, brush1,
                new RectangleF(new PointF(x1, y), size), format2);

            // Adjust the position for the next line
            y = y + size.Height + font3.Height * 0.75f;

            // Build a StringBuilder to concatenate the remaining lines of content
            StringBuilder buff = new StringBuilder();
            for (int i = 3; i < lines.Length; i++)
            {
                buff.AppendLine(lines[i]);
            }
            content = buff.ToString();

            // Create a PdfStringLayouter for arranging the remaining content
            // Instantiate a PdfStringLayouter object for arranging the remaining content
            PdfStringLayouter textLayouter = new PdfStringLayouter();

            // Use the textLayouter to layout the remaining content on the page with the specified font, format, and size
            PdfStringLayoutResult result = textLayouter.Layout(content, font3, format2, new SizeF(width, float.MaxValue));

            // Iterate through each line in the layout result
            foreach (LineInfo line in result.Lines)
            {
                // Check if the line is the first line of a paragraph
                if ((line.LineType & LineType.FirstParagraphLine) == LineType.FirstParagraphLine)
                {
                    // Adjust the vertical position to create some spacing before the paragraph
                    y = y + font3.Height * 0.75f;
                }

                // Check if the current position exceeds the available space on the current page
                if (y > (page.Canvas.ClientSize.Height - margin.Bottom - result.LineHeight))
                {
                    // Add a new page to the section
                    page = section.Pages.Add();

                    // Call a method to draw the header and footer on the newly added page without cover information
                    DrawPageHeaderAndFooter(page, margin, false);

                    // Reset the vertical position to the top margin
                    y = margin.Top;
                }

                // Draw the line of text on the page canvas using the specified font, brush, position, and format
                page.Canvas.DrawString(line.Text, font3, brush1, x, y, format2);

                // Adjust the vertical position for the next line
                y = y + result.LineHeight;
            }
        }

        private void DrawPageHeaderAndFooter(PdfPageBase page, PdfMargins margin, bool isCover)
        {
            // Set the transparency of the canvas to 0.5 (50% opacity)
            page.Canvas.SetTransparency(0.5f);

            // Load the header and footer images from files
            PdfImage headerImage = PdfImage.FromFile(@"..\..\..\..\..\..\Data\Header.png");
            PdfImage footerImage = PdfImage.FromFile(@"..\..\..\..\..\..\Data\Footer.png");

            // Draw the header image at the top-left corner of the page
            page.Canvas.DrawImage(headerImage, new PointF(0, 0));

            // Draw the footer image at the bottom-left corner of the page
            page.Canvas.DrawImage(footerImage, new PointF(0, page.Canvas.ClientSize.Height - footerImage.PhysicalDimension.Height));

            // Check if it is a cover page
            if (isCover)
            {
                // Set the transparency of the canvas to 1 (100% opacity) for the cover page
                page.Canvas.SetTransparency(1);
                return;
            }

            // Set up brush, pen, font, format, and spacing for the header and footer text
            PdfBrush brush = PdfBrushes.Black;
            PdfPen pen = new PdfPen(brush, 0.75f);
            PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Arial", 9f, FontStyle.Italic), true);
            PdfStringFormat format = new PdfStringFormat(PdfTextAlignment.Right);
            format.MeasureTrailingSpaces = true;
            float space = font.Height * 0.75f;

            // Set up initial positions and dimensions
            float x = margin.Left;
            float width = page.Canvas.ClientSize.Width - margin.Left - margin.Right;
            float y = margin.Top - space;

            // Draw a horizontal line below the header
            page.Canvas.DrawLine(pen, x, y, x + width, y);

            // Adjust the position for the header text
            y = y - 1 - font.Height;

            // Draw the header text "Demo of Spire.Pdf" on the right side of the page
            page.Canvas.DrawString("Demo of Spire.Pdf", font, brush, x + width, y, format);

            // Set the transparency of the canvas to 1 (100% opacity)
            page.Canvas.SetTransparency(1);
        }

        private void DrawPageNumber(PdfSection section, PdfMargins margin, int startNumber, int pageCount)
        {
            // Iterate through each page in the section
            foreach (PdfPageBase page in section.Pages)
            {
                // Set the transparency of the canvas to 0.5 (50% opacity)
                page.Canvas.SetTransparency(0.5f);

                // Set up brush, pen, font, format, and spacing for the page number text
                PdfBrush brush = PdfBrushes.Black;
                PdfPen pen = new PdfPen(brush, 0.75f);
                PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Arial", 9f, FontStyle.Italic), true);
                PdfStringFormat format = new PdfStringFormat(PdfTextAlignment.Right);
                format.MeasureTrailingSpaces = true;
                float space = font.Height * 0.75f;

                // Set up initial positions and dimensions
                float x = margin.Left;
                float width = page.Canvas.ClientSize.Width - margin.Left - margin.Right;
                float y = page.Canvas.ClientSize.Height - margin.Bottom + space;

                // Draw a horizontal line above the footer
                page.Canvas.DrawLine(pen, x, y, x + width, y);

                // Adjust the position for the page number text
                y = y + 1;

                // Create the page number label using the format "startNumber of pageCount"
                String numberLabel = String.Format("{0} of {1}", startNumber++, pageCount);

                // Draw the page number label on the right side of the page
                page.Canvas.DrawString(numberLabel, font, brush, x + width, y, format);

                // Set the transparency of the canvas to 1 (100% opacity)
                page.Canvas.SetTransparency(1);
            }
        }

        private void PDFDocumentViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }

    }
}

File: VB/14_Page/Pagination.vb
Imports System.Text
Imports Spire.Pdf
Imports Spire.Pdf.Graphics

Namespace Pagination
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			' Create a new PDF document
			Dim doc As New PdfDocument()

			' Set the margin for the document
			Dim unitCvtr As New PdfUnitConvertor()
			Dim margin As New PdfMargins()
			margin.Top = unitCvtr.ConvertUnits(2.54F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)
			margin.Bottom = margin.Top
			margin.Left = unitCvtr.ConvertUnits(3.17F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)
			margin.Right = margin.Left

			' Draw the cover page using the specified margin
			DrawCover(doc, doc.Sections.Add(), margin)

			' Draw the content pages using the specified margin
			DrawContent(doc, doc.Sections.Add(), margin)

			' Draw the page numbers on the second section of the document
			DrawPageNumber(doc.Sections(1), margin, 1, doc.Sections(1).Pages.Count)

			' Save the document to a file
			doc.SaveToFile("Pagination.pdf")

			' Close the document
			doc.Close()

			' Launch the Pdf file
			PDFDocumentViewer("Pagination.pdf")
		End Sub

		Private Sub DrawCover(ByVal pdf As PdfDocument, ByVal section As PdfSection, ByVal margin As PdfMargins)
			' Set the page size of the section to A4
			section.PageSettings.Size = PdfPageSize.A4

			' Set the margins of the section to 0 on all sides
			section.PageSettings.Margins.All = 0

			' Add a new page to the section
			Dim page As PdfPageBase = section.Pages.Add()

			' Create a new instance of PdfPageLabels for the document
			pdf.PageLabels = New PdfPageLabels()

			' Add page labels starting from index 0 with lowercase Roman numerals in the format "cover "
			pdf.PageLabels.AddRange(0, PdfPageLabels.Lowercase_Roman_Numerals_Style, "cover ")

			' Draw the page header and footer on the page using the specified margin
			DrawPageHeaderAndFooter(page, margin, True)

			' Create brushes and fonts for drawing text and images
			Dim brush1 As PdfBrush = PdfBrushes.LightGray
			Dim brush2 As PdfBrush = PdfBrushes.Blue
			Dim font1 As New PdfTrueTypeFont(New Font("Arial", 8.0F))
			Dim format As New PdfStringFormat()
			format.MeasureTrailingSpaces = True

			' Specify the text content for the information section
			Dim text1 As String = "(All text and picture from "
			Dim text2 As String = "Wikipedia"
			Dim text3 As String = ", the free encyclopedia)"

			' Initialize the x and y coordinates for drawing
			Dim x As Single = 0, y As Single = 10

			' Increment the x and y coordinates by the left margin
			x = x + margin.Left
			y = y + margin.Top

			' Draw the first part of the information text
			page.Canvas.DrawString(text1, font1, brush1, x, y, format)

			' Update the x-coordinate for the next part of the text
			x = x + font1.MeasureString(text1, format).Width

			' Draw the second part of the information text
			page.Canvas.DrawString(text2, font1, brush2, x, y, format)

			' Update the x-coordinate for the next part of the text
			x = x + font1.MeasureString(text2, format).Width

			' Draw the third part of the information text
			page.Canvas.DrawString(text3, font1, brush1, x, y, format)

			' Create brushes and pens for drawing an image with description
			Dim brush3 As PdfBrush = PdfBrushes.Black
			Dim brush4 As PdfBrush = New PdfSolidBrush(New PdfRGBColor(&HF9, &HF9, &HF9))
			Dim image As PdfImage = PdfImage.FromFile("..\..\..\..\..\..\Data\SciencePersonificationBoston.jpg")
			Dim text As String = My.Resources.ImageDescription
			Dim r As Single = image.PhysicalDimension.Height / image.Height
			Dim pen As New PdfPen(brush1, r)
			Dim size As SizeF = font1.MeasureString(text, image.PhysicalDimension.Width - 2)

			' Create a template for the image with description
			Dim template As New PdfTemplate(image.PhysicalDimension.Width + 4 * r + 4, image.PhysicalDimension.Height + 4 * r + 7 + size.Height)

			' Draw a rectangle border around the template
			template.Graphics.DrawRectangle(pen, brush4, 0, 0, template.Width, template.Height)

			' Set the initial coordinates for drawing the image and description within the template
			y = r + 2
			x = y

			' Draw a rectangle around the image within the template
			template.Graphics.DrawRectangle(brush1, x, y, image.PhysicalDimension.Width + 2 * r, image.PhysicalDimension.Height + 2 * r)

			' Update the coordinates for drawing the image within the rectangle
			y = x + r
			x = y

			' Draw the image within the rectangle
			template.Graphics.DrawImage(image, x, y)

			' Adjust the coordinates for drawing the description below the image
			x = x - 1
			y = y + image.PhysicalDimension.Height + r + 2

			' Draw the description text within a rectangle below the image
			template.Graphics.DrawString(text, font1, brush3, New RectangleF(New PointF(x, y), size))

			' Calculate the position to draw the template on the page's canvas
			Dim x1 As Single = (page.Canvas.ClientSize.Width - template.Width) / 2
			Dim y1 As Single = (page.Canvas.ClientSize.Height - margin.Top - margin.Bottom) * (1 - 0.618F) - template.Height \ 2 + margin.Top

			' Draw the template on the page's canvas at the specified position
			template.Draw(page.Canvas, x1, y1)

			' Set the alignment for drawing the title text
			format.Alignment = PdfTextAlignment.Center

			' Create font2 for drawing text
			Dim font2 As New PdfTrueTypeFont(New Font("Arial", 24.0F, FontStyle.Bold))
			x = page.Canvas.ClientSize.Width / 2
			y = y1 + template.Height + 10

			' Draw text
			page.Canvas.DrawString(My.Resources.Title, font2, brush3, x, y, format)
		End Sub

		Private Sub DrawContent(ByVal pdf As PdfDocument, ByVal section As PdfSection, ByVal margin As PdfMargins)
			' Set the page size of the section to A4
			section.PageSettings.Size = PdfPageSize.A4

			' Set the margins of the section to 0 on all sides
			section.PageSettings.Margins.All = 0

			' Add a new page to the section
			Dim page As PdfPageBase = section.Pages.Add()

			' Add page labels starting from index 1 with decimal Arabic numerals in the format "page "
			pdf.PageLabels.AddRange(1, PdfPageLabels.Decimal_Arabic_Numerals_Style, "page ")

			' Draw the page header and footer on the page using the specified margin
			DrawPageHeaderAndFooter(page, margin, False)

			' Set the initial coordinates and width for drawing text
			Dim x As Single = margin.Left
			Dim y As Single = margin.Top + 8
			Dim width As Single = page.Canvas.ClientSize.Width - margin.Left - margin.Right

			' Create a font and brushes for drawing the title
			Dim font1 As New PdfTrueTypeFont(New Font("Arial", 16.0F), True)
			Dim brush1 As PdfBrush = PdfBrushes.Black

			' Create a pen for drawing a horizontal line
			Dim pen1 As New PdfPen(brush1, 0.75F)

			' Draw the title text
			page.Canvas.DrawString(My.Resources.Title, font1, brush1, x, y)

			' Update the y-coordinate for the next elements
			y = y + font1.MeasureString(My.Resources.Title).Height + 6

			' Draw a horizontal line below the title
			page.Canvas.DrawLine(pen1, x, y, page.Canvas.ClientSize.Width - margin.Right, y)

			' Update the y-coordinate for the next elements
			y = y + 1.75F

			' Get the content from resources
			Dim content As String = My.Resources.Content

			' Split the content into lines
			Dim lines() As String = content.Split(New Char() {ControlChars.Cr, ControlChars.Lf}, StringSplitOptions.RemoveEmptyEntries)

			' Create a font and string format for the content text
			Dim font2 As New PdfTrueTypeFont(New Font("Arial", 10.0F, FontStyle.Italic), True)
			Dim format1 As New PdfStringFormat()
			format1.MeasureTrailingSpaces = True
			format1.LineSpacing = font2.Height * 1.5F
			format1.ParagraphIndent = font2.MeasureString(vbTab, format1).Width

			' Update the y-coordinate for the next elements
			y = y + font2.Height * 0.5F

			' Measure and draw the first line of content text
			Dim size As SizeF = font2.MeasureString(lines(0), width, format1)
			page.Canvas.DrawString(lines(0), font2, brush1, New RectangleF(New PointF(x, y), size), format1)

			' Update the y-coordinate for the next elements
			y = y + size.Height

			' Create a font and string format for subsequent lines of content text
			Dim font3 As New PdfTrueTypeFont(New Font("Arial", 10.0F), True)
			Dim format2 As New PdfStringFormat()
			format2.LineSpacing = font3.Height * 1.4F
			format2.MeasureTrailingSpaces = True

			' Measure and draw the second line of content text
			size = font3.MeasureString(lines(1), width, format2)
			page.Canvas.DrawString(lines(1), font3, brush1, New RectangleF(New PointF(x, y), size), format2)

			' Update the y-coordinate for the next elements
			y = y + size.Height

			' Increment the y-coordinate and calculate the indentation for the third line of content text
			y = y + font3.Height * 0.75F
			Dim indent As Single = font3.MeasureString(vbTab & vbTab, format2).Width
			Dim x1 As Single = x + indent

			' Measure and draw the third line of content text with indentation
			size = font3.MeasureString(lines(2), width - indent, format2)
			page.Canvas.DrawString(lines(2), font3, brush1, New RectangleF(New PointF(x1, y), size), format2)

			' Update the y-coordinate for the next elements
			y = y + size.Height + font3.Height * 0.75F

			' Create a StringBuilder to store the remaining lines of content text
			Dim buff As New StringBuilder()
			For i As Integer = 3 To lines.Length - 1
				buff.AppendLine(lines(i))
			Next i
			content = buff.ToString()

			' Create a PdfStringLayouter to lay out the remaining content text
			Dim textLayouter As New PdfStringLayouter()
			Dim result As PdfStringLayoutResult = textLayouter.Layout(content, font3, format2, New SizeF(width, Single.MaxValue))

			' Iterate through each line in the layout result and draw it on the page
			For Each line As LineInfo In result.Lines
				If (line.LineType And LineType.FirstParagraphLine) = LineType.FirstParagraphLine Then
					y = y + font3.Height * 0.75F
				End If
				If y > (page.Canvas.ClientSize.Height - margin.Bottom - result.LineHeight) Then
					page = section.Pages.Add()
					DrawPageHeaderAndFooter(page, margin, False)
					y = margin.Top
				End If
				page.Canvas.DrawString(line.Text, font3, brush1, x, y, format2)
				y = y + result.LineHeight
			Next line
		End Sub

		Private Sub DrawPageHeaderAndFooter(ByVal page As PdfPageBase, ByVal margin As PdfMargins, ByVal isCover As Boolean)
			' Load the header and footer images from file
			Dim headerImage As PdfImage = PdfImage.FromFile("..\..\..\..\..\..\Data\Header.png")
			Dim footerImage As PdfImage = PdfImage.FromFile("..\..\..\..\..\..\Data\Footer.png")

			' Draw the header image at the top-left corner of the page's canvas
			page.Canvas.DrawImage(headerImage, New PointF(0, 0))

			' Draw the footer image at the bottom-left corner of the page's canvas
			page.Canvas.DrawImage(footerImage, New PointF(0, page.Canvas.ClientSize.Height - footerImage.PhysicalDimension.Height))

			' If it is a cover page, set transparency and return
			If isCover Then
				page.Canvas.SetTransparency(1)
				Return
			End If

			' Create brushes, pen, font, and format for drawing text and lines
			Dim brush As PdfBrush = PdfBrushes.Black
			Dim pen As New PdfPen(brush, 0.75F)
			Dim font As New PdfTrueTypeFont(New Font("Arial", 9.0F, FontStyle.Italic), True)
			Dim format As New PdfStringFormat(PdfTextAlignment.Right)
			format.MeasureTrailingSpaces = True

			' Calculate spacing and coordinates for drawing elements
			Dim space As Single = font.Height * 0.75F
			Dim x As Single = margin.Left
			Dim width As Single = page.Canvas.ClientSize.Width - margin.Left - margin.Right
			Dim y As Single = margin.Top - space

			' Draw a horizontal line at the top of the page
			page.Canvas.DrawLine(pen, x, y, x + width, y)

			' Update the y-coordinate for drawing the text
			y = y - 1 - font.Height

			' Draw the text "Demo of Spire.Pdf" on the right side of the page
			page.Canvas.DrawString("Demo of Spire.Pdf", font, brush, x + width, y, format)

			' Set transparency to fully opaque (1)
			page.Canvas.SetTransparency(1)
		End Sub

		Private Sub DrawPageNumber(ByVal section As PdfSection, ByVal margin As PdfMargins, ByVal startNumber As Integer, ByVal pageCount As Integer)
			' Iterate through each page in the section
			For Each page As PdfPageBase In section.Pages
				' Set transparency to 0.5 (semi-transparent)
				page.Canvas.SetTransparency(0.5F)

				' Create brushes, pen, font, and format for drawing text and lines
				Dim brush As PdfBrush = PdfBrushes.Black
				Dim pen As New PdfPen(brush, 0.75F)
				Dim font As New PdfTrueTypeFont(New Font("Arial", 9.0F, FontStyle.Italic), True)
				Dim format As New PdfStringFormat(PdfTextAlignment.Right)
				format.MeasureTrailingSpaces = True

				' Calculate spacing and coordinates for drawing elements
				Dim space As Single = font.Height * 0.75F
				Dim x As Single = margin.Left
				Dim width As Single = page.Canvas.ClientSize.Width - margin.Left - margin.Right
				Dim y As Single = page.Canvas.ClientSize.Height - margin.Bottom + space

				' Draw a horizontal line at the bottom of the page
				page.Canvas.DrawLine(pen, x, y, x + width, y)

				' Update the y-coordinate for drawing the text
				y = y + 1

				' Generate the number label for the page
				Dim numberLabel As String = String.Format("{0} of {1}", startNumber, pageCount)

				' Increment the startNumber for the next page
				startNumber += 1

				' Draw the number label on the right side of the page
				page.Canvas.DrawString(numberLabel, font, brush, x + width, y, format)

				' Set transparency to fully opaque (1)
				page.Canvas.SetTransparency(1)
			Next page
		End Sub

		Private Sub PDFDocumentViewer(ByVal fileName As String)
			Try
				Process.Start(fileName)
			Catch
			End Try
		End Sub

	End Class
End Namespace
